pulsar/src/main/java/com/flipkart/varadhi/pulsar/entities/PulsarProducer.java

```diff
@@ -2,19 +2,123 @@
+                // failures are unknown.
+                produceFuture.completeExceptionally(throwable);
+            }
+        });
+        return produceFuture;
+    }
+
+
+    private org.apache.pulsar.client.api.Producer<byte[]> getProducer(
+            PulsarStorageTopic topic, ProducerOptions options, String hostname
+    )
+            throws PulsarClientException {
+        Map<String, Object> producerConfig = getProducerConfig(topic, options, hostname);
+        return clientProvider.getPulsarClient().newProducer().loadConf(producerConfig).create();
+    }
+
+    private Map<String, Object> getProducerConfig(PulsarStorageTopic topic, ProducerOptions options, String hostName) {
+
+        // System Configured::
+        // sendTimeoutMs
+        // blockIfQueueFull
+        // compressionType
+        // batchingEnabled
+        // batchingMaxPublishDelay
+
+        // Fixed::
+        // name = topic + hostname
+        // topic = topic name
+        // access mode -- shared.
+        //
+        // Default Values::
+        // enableChunking: false
+        // autoUpdatePartitions, autoUpdatePartitionsInterval
+        //
+        // Calculated:: Topic capacity dependent, (primitive calculation by Varadhi).
+        // maxPendingMessages, maxPendingMessagesAcrossPartitions
+        // batchingMaxPublishDelay, batchingMaxMessages, batchingMaxBytes,
+
+        Map<String, Object> producerConfig = options.asConfigMap();
+        producerConfig.put("topicName", topic.getName());
+        producerConfig.put("producerName", getProducerName(topic.getName(), hostName));
+        producerConfig.put("accessMode", ProducerAccessMode.Shared);
+
+        int topicMaxQps = topic.getMaxQPS();
+        int maxPendingMessages = getMaxPendingMessages(topicMaxQps);
+        int batchingMaxMessages = getBatchMaxMessages(topicMaxQps, options.getBatchingMaxPublishDelayMs());
+        producerConfig.put("maxPendingMessages", maxPendingMessages);
+        // maxPendingMessages and maxPendingMessagesAcrossPartitions are kept same assuming worst case.
+        producerConfig.put("maxPendingMessagesAcrossPartitions", maxPendingMessages);
+        producerConfig.put("batchingMaxMessages", batchingMaxMessages);
+        producerConfig.put("batchingMaxBytes", batchingMaxMessages * getBatchingMaxBytes(batchingMaxMessages, topic));
+        return producerConfig;
     }
+
+    private String getProducerName(String topicName, String hostName) {
+        String topicSuffix = topicName.split("/")[4];
+        return String.format("%s.%s", topicSuffix, hostName);
+    }
+
+    private int getMaxPendingMessages(int topicMaxQps) {
+        // Assumption: Don't allow queue to build more than 1 second worth of messages.
+        // with bound [min, max]
+        // TODO:: This impacts memory so Discuss and close it.
+        return Math.min(MAX_PENDING_MESSAGES, Math.max(MIN_PENDING_MESSAGES, topicMaxQps));
```

Comment:
I actually can;t comment right now. Anil did some work here. But here the even the topicQPS will be distributed across pods. So it is "technically" not correct to use topicMaxQps here.

I haven't experimented with produce configs like these, to comment. Dont have full understanding still.

Reply:
sure. This needs to be tuned based on benchmark and observations. As of now approx calculation (assuming worst case of all messages to the same producer) to allow messages worth 1 sec with both lower/upper bound.